libherokubuildpack/src/output/background_dots.rs

```diff
@@ -0,0 +1,161 @@
+            write!(buffer, "{}", style::background_timer_tick()).expect("Internal error");
+            buffer.flush().expect("Internal error");
+
+            if receiver.recv_timeout(print_interval).is_ok() {
+                break;
+            }
+        }
+
+        write!(buffer, "{}", style::background_timer_end()).expect("Internal error");
+        buffer.flush().expect("Internal error");
+
+        buffer
+    });
+
+    DotPrintState::new(join_handle, sender)
+}
+
+#[derive(Debug)]
+pub(crate) struct DotPrintState<W>
+where
+    W: Write + Debug,
+{
+    join_handle: Option<JoinHandle<W>>,
+    stop_signal: Sender<()>,
+}
+
+impl<W> Drop for DotPrintState<W>
+where
+    W: Write + Debug,
+{
+    fn drop(&mut self) {
+        if let Some(join_handle) = self.join_handle.take() {
+            let _ = self.stop_signal.send(());
+            let _ = join_handle.join();
+        }
+    }
+}
+
+impl<W> DotPrintState<W>
+where
+    W: Write + Debug,
+{
+    fn new(join_handle: JoinHandle<W>, sender: Sender<()>) -> Self {
+        DotPrintState {
+            join_handle: Some(join_handle),
+            stop_signal: sender,
+        }
+    }
+
+    /// The only thing a consumer can do is stop the dots and receive
+    /// the original buffer.
+    ///
+    /// # Panics
+    ///
+    /// This code can panic if it encounters an unexpedcted internal state.
+    /// If that happens it means there is an internal bug in this logic.
+    /// To avoid a panic, developers modifying this file must obey the following
+    /// rules:
+    ///
+    /// - Always consume the struct when accessing the Option value outside of Drop
+    /// - Never construct this struct with a `None` option value.
+    ///
+    /// This `Option` wrapping is needed to support a implementing Drop to
+    /// ensure the printing is stopped. When a struct implements drop it cannot
+    /// remove it's internal state due to E0509:
+    ///
+    /// <https://github.com/rust-lang/rust/blob/27d8a577138c0d319a572cd1a464c2b755e577de/compiler/rustc_error_codes/src/error_codes/E0509.md>
+    ///
+    /// The workaround is to never allow invalid internal state by replacing the
+    /// inner value with a `None` when removing it. We don't want to expose this
+    /// implementation detail to the user, so instead we accept the panic, ensure
+    /// the code is exercised under test, and exhaustively document why this panic
+    /// exists and how developers working with this code can maintain safety
+    #[allow(clippy::panic_in_result_fn)]
+    pub(crate) fn stop(mut self) -> std::thread::Result<W> {
+        // Ignore if the channel is closed, likely means the thread died which
+        // we want in this case.
+        match self.join_handle.take() {
+            Some(join_handle) => {
+                let _ = self.stop_signal.send(());
+                join_handle.join()
+            }
+            None => panic!("Internal error: Dot print internal state should never be None"),
```

Comment:
Another alternative is to use something like 

- https://crates.io/crates/consume_on_drop

Which uses `ManuallyDrop` under  the hood https://github.com/markcsaving/consume_on_drop/blob/2da5decaba986313e8db1d0693b70058dd1757bd/src/lib.rs#L35-L37

In a similar territory I also found https://crates.io/crates/dispose which allows us to use the consuming interface for drop, though I'm not quite clear if it also allows you to manually get the value. This also uses ManuallyDrop https://github.com/ray-kast/dispose/blob/bd3675c164f1135ac39d35925c9d16a4497776be/src/dispose/disposable.rs#L20. 

The implementation is small enough we could re-write something similar without pulling in a dependency.

The benefit is we would get rid of this theoretically possible panic. The downside is that we trade it for an unsafe block. I'm unsure how exactly ManuallyDrop interacts with Drop i.e. if we ManuallyDrop::take the inner, then Drop is called on the outer struct, I don't know exactly how those play together.

I think this `Option` is good enough for our use case but I wanted to note that I looked into ManuallyDrop to understand all my options.

Reply:
We talked about it. The panic is still not ideal, but it's not much code and seems okay. I added a debug_assert and moved it into it's own module for stronger guarantees only the public interfaces were being used.